tiledb/sm/query/dimension_label/dimension_label_query.h

```diff
@@ -0,0 +1,231 @@
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Classes for query a dimension label.
+ */
+
+#ifndef TILEDB_DIMENSION_LABEL_QUERY_H
+#define TILEDB_DIMENSION_LABEL_QUERY_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/type/range/range.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+class DimensionLabel;
+class Query;
+class QueryBuffer;
+class StorageManager;
+class Subarray;
+
+class DimensionLabelQuery {
+ public:
+  /** Default constructor. */
+  DimensionLabelQuery() = default;
+
+  /**
+   * General constructor.
+   *
+   * @param storage_manager Storage manager object.
+   * @param dimension_label Opened dimension label for the query.
+   * @param add_indexed_query If ``true``, create a query on the indexed array.
+   * @param add_labelled_query If ``true``, create a query on the labelled
+   *     array.
+   * @param fragment_name Optional fragment name for writing fragments.
+   */
+  DimensionLabelQuery(
+      StorageManager* storage_manager,
+      DimensionLabel* dimension_label,
+      bool add_indexed_query,
+      bool add_labelled_query,
+      optional<std::string> fragment_name = nullopt);
+
+  /** Destructor. */
+  virtual ~DimensionLabelQuery() = default;
+
+  /** Disable copy and move. */
+  DISABLE_COPY_AND_COPY_ASSIGN(DimensionLabelQuery);
+  DISABLE_MOVE_AND_MOVE_ASSIGN(DimensionLabelQuery);
+
+  /** Cancel both queries if they exist. */
+  void cancel();
+
+  /** Returns ``true`` if the query status for both queries is completed. */
+  bool completed() const;
+
+  /** Finalizes both queries if they exist. */
+  void finalize();
+
+  /** Processes both queries if they exist. */
+  void process();
+
+ protected:
+  /** Query on the dimension label indexed array. */
+  tdb_unique_ptr<Query> indexed_array_query{nullptr};
```

Comment:
They could be. Not sure it's worth changing since they will be replaced with a single query that is always initialized once we get all the dimension label readers/writers in. Let me know if you feel strongly about this, and I will update it.

Reply:
I added a note to the top of the file about the temporary design.